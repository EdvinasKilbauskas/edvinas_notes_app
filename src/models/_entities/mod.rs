//! `SeaORM` Entity, @generated by sea-orm-codegen 1.0.0

pub mod prelude;
pub mod note_shares;
pub mod notes;
pub mod users;
