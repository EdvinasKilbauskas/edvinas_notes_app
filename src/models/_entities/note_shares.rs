use sea_orm::entity::prelude::*;
use serde::{ Deserialize, Serialize };

#[derive(Clone, Debug, PartialEq, DeriveEntityModel, Eq, Serialize, Deserialize)]
#[sea_orm(table_name = "note_shares")]
pub struct Model {
    #[sea_orm(primary_key)]
    pub id: i32,
    pub note_id: i32,
    pub shared_with_user_id: i32,
}

#[derive(Copy, Clone, Debug, EnumIter, DeriveRelation)]
pub enum Relation {
    #[sea_orm(
        belongs_to = "super::notes::Entity",
        from = "Column::NoteId",
        to = "super::notes::Column::Id"
    )]
    Note,
    #[sea_orm(
        belongs_to = "super::users::Entity",
        from = "Column::SharedWithUserId",
        to = "super::users::Column::Id"
    )]
    SharedWithUser,
}

impl Related<super::notes::Entity> for Entity {
    fn to() -> RelationDef {
        Relation::Note.def()
    }
}

impl Related<super::users::Entity> for Entity {
    fn to() -> RelationDef {
        Relation::SharedWithUser.def()
    }
}

impl ActiveModelBehavior for ActiveModel {}
