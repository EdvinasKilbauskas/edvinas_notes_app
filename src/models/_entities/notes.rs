use sea_orm::entity::prelude::*;
use serde::{ Deserialize, Serialize };

#[derive(Clone, Debug, PartialEq, DeriveEntityModel, Eq, Serialize, Deserialize)]
#[sea_orm(table_name = "notes")]
pub struct Model {
    pub created_at: DateTimeWithTimeZone,
    pub updated_at: DateTimeWithTimeZone,
    #[sea_orm(primary_key)]
    pub id: i32,
    pub title: Option<String>,
    pub content: Option<String>,
    pub user_id: i32, // Add this line
}

#[derive(Copy, Clone, Debug, EnumIter, DeriveRelation)]
pub enum Relation {
    #[sea_orm(
        belongs_to = "super::users::Entity",
        from = "Column::UserId",
        to = "super::users::Column::Id"
    )]
    User,
    #[sea_orm(has_many = "super::note_shares::Entity")]
    NoteShares,
}

impl Related<super::users::Entity> for Entity {
    fn to() -> RelationDef {
        Relation::User.def()
    }
}

impl Related<super::note_shares::Entity> for Entity {
    fn to() -> RelationDef {
        Relation::NoteShares.def()
    }
}
