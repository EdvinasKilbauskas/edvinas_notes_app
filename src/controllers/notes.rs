#![allow(clippy::missing_errors_doc)]
#![allow(clippy::unnecessary_struct_initialization)]
#![allow(clippy::unused_async)]
use axum::debug_handler;
use loco_rs::prelude::*;
use serde::{Deserialize, Serialize};
use crate::models::_entities::notes::{ActiveModel, Column, Entity, Model};
use crate::models::_entities::users;
use crate::models::_entities::note_shares::{self, ActiveModel as NoteShareActiveModel};
use sea_orm::*;

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct ShareNoteParams {
    pub shared_with_user_id: i32,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct SharedNoteResponse {
    pub id: i32,
    pub title: Option<String>,
    pub content: Option<String>,
    pub shared_by_user_id: i32,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct Params {
    pub title: Option<String>,
    pub content: Option<String>,
}

impl Params {
    fn update(&self, item: &mut ActiveModel) {
        item.title = Set(self.title.clone());
        item.content = Set(self.content.clone());
    }
}

use sea_orm::{ColumnTrait, EntityTrait, QueryFilter, QuerySelect};

async fn load_item(ctx: &AppContext, id: i32, user_id: i32) -> Result<Model> {
    let item = Entity::find_by_id(id)
        .filter(
            crate::models::_entities::notes::Column::UserId.eq(user_id)
                .or(
                    crate::models::_entities::note_shares::Column::NoteId.eq(id)
                        .and(crate::models::_entities::note_shares::Column::SharedWithUserId.eq(user_id))
                )
        )
        .join(
            JoinType::LeftJoin,
            crate::models::_entities::notes::Relation::NoteShares.def()
        )
        .one(&ctx.db)
        .await?;
    
    item.ok_or_else(|| Error::NotFound)
}


#[debug_handler]
pub async fn list(auth: auth::JWT, State(ctx): State<AppContext>) -> Result<Response> {
    let user = users::Model::find_by_pid(&ctx.db, &auth.claims.pid).await?;
    
    let notes = Entity::find()
        .filter(
            Condition::any()
                .add(crate::models::_entities::notes::Column::UserId.eq(user.id))
                .add(
                    crate::models::_entities::notes::Column::Id.in_subquery(
                        note_shares::Entity::find()
                            .select_only()
                            .column(note_shares::Column::NoteId)
                            .filter(note_shares::Column::SharedWithUserId.eq(user.id))
                            .into_query()
                    )
                )
        )
        .all(&ctx.db)
        .await?;
    
    format::json(notes)
}

#[debug_handler]
pub async fn add(auth: auth::JWT, State(ctx): State<AppContext>, Json(params): Json<Params>) -> Result<Response> {
    let user = users::Model::find_by_pid(&ctx.db, &auth.claims.pid).await?;
    let mut item = ActiveModel {
        user_id: Set(user.id),
        
        ..Default::default()
    };
    params.update(&mut item);
    let item = item.insert(&ctx.db).await?;
    format::json(item)
}

#[debug_handler]
pub async fn update(
    auth: auth::JWT,
    Path(id): Path<i32>,
    State(ctx): State<AppContext>,
    Json(params): Json<Params>,
) -> Result<Response> {
    let user = users::Model::find_by_pid(&ctx.db, &auth.claims.pid).await?;
    let item = Entity::find_by_id(id)
        .filter(crate::models::_entities::notes::Column::UserId.eq(user.id))
        .one(&ctx.db)
        .await?
        .ok_or_else(|| Error::NotFound)?;
    let mut item = item.into_active_model();
    params.update(&mut item);
    let item = item.update(&ctx.db).await?;
    format::json(item)
}

#[debug_handler]
pub async fn remove(auth: auth::JWT, Path(id): Path<i32>, State(ctx): State<AppContext>) -> Result<Response> {
    let user = users::Model::find_by_pid(&ctx.db, &auth.claims.pid).await?;
    let item = Entity::find_by_id(id)
        .filter(crate::models::_entities::notes::Column::UserId.eq(user.id))
        .one(&ctx.db)
        .await?
        .ok_or_else(|| Error::NotFound)?;
    item.delete(&ctx.db).await?;
    format::empty()
}

#[debug_handler]
pub async fn get_notes_shared_by_me(auth: auth::JWT, State(ctx): State<AppContext>) -> Result<Response> {
    let user = users::Model::find_by_pid(&ctx.db, &auth.claims.pid).await?;
    
    let shared_notes = Entity::find()
        .filter(crate::models::_entities::notes::Column::UserId.eq(user.id))
        .join(JoinType::InnerJoin, crate::models::_entities::notes::Relation::NoteShares.def())
        .group_by(crate::models::_entities::notes::Column::Id)
        .all(&ctx.db)
        .await?;
    
    format::json(shared_notes)
}

#[debug_handler]
pub async fn get_one(auth: auth::JWT, Path(id): Path<i32>, State(ctx): State<AppContext>) -> Result<Response> {
    let user = users::Model::find_by_pid(&ctx.db, &auth.claims.pid).await?;
    format::json(load_item(&ctx, id, user.id).await?)
}

#[debug_handler]
pub async fn share_note(
    auth: auth::JWT,
    Path(note_id): Path<i32>,
    State(ctx): State<AppContext>,
    Json(params): Json<ShareNoteParams>,
) -> Result<Response> {
    let user = users::Model::find_by_pid(&ctx.db, &auth.claims.pid).await?;
    let note = load_item(&ctx, note_id, user.id).await?;
    
    let share = NoteShareActiveModel {
        note_id: Set(note.id),
        shared_with_user_id: Set(params.shared_with_user_id),
        ..Default::default()
    };
    let share = share.insert(&ctx.db).await?;
    
    format::json(share)
}



#[debug_handler]
pub async fn share_all_notes(
    auth: auth::JWT,
    State(ctx): State<AppContext>,
    Json(params): Json<ShareNoteParams>,
) -> Result<Response> {
    let user = users::Model::find_by_pid(&ctx.db, &auth.claims.pid).await?;
    
    // Get all notes of the current user
    let user_notes = Entity::find()
        .filter(Column::UserId.eq(user.id))
        .all(&ctx.db)
        .await?;

    // Share each note with the specified user
    for note in user_notes {
        let share = note_shares::ActiveModel {
            note_id: Set(note.id),
            shared_with_user_id: Set(params.shared_with_user_id),
            ..Default::default()
        };
        share.insert(&ctx.db).await?;
    }

    format::json(serde_json::json!({
        "message": "All notes have been shared successfully"
    }))
}

#[debug_handler]
pub async fn get_shared_notes(auth: auth::JWT, State(ctx): State<AppContext>) -> Result<Response> {
    let user = users::Model::find_by_pid(&ctx.db, &auth.claims.pid).await?;
    
    let shared_notes = Entity::find()
        .join(JoinType::InnerJoin, crate::models::_entities::notes::Relation::NoteShares.def())
        .filter(crate::models::_entities::note_shares::Column::SharedWithUserId.eq(user.id))
        .group_by(crate::models::_entities::notes::Column::Id)
        .all(&ctx.db)
        .await?;
    
    format::json(shared_notes)
}

pub fn routes() -> Routes {
    Routes::new()
        .prefix("notes")
        .add("/", get(list))
        .add("/", post(add))
        .add("/:id", get(get_one))
        .add("/:id", delete(remove))
        .add("/:id", post(update))
        .add("/:id/share", post(share_note))
        .add("/shared", get(get_shared_notes))
        .add("/shared-by-me", get(get_notes_shared_by_me))
        .add("/share-all", post(share_all_notes))
}