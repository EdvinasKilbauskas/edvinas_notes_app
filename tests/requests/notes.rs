use axum::http::{ HeaderName, HeaderValue };
use insta::{ assert_debug_snapshot, with_settings };
use loco_rs::testing;
use edvinas_notes_app::{ app::App, models::_entities::notes::Entity };
use edvinas_notes_app::models::_entities::note_shares;
use edvinas_notes_app::models::_entities::notes;
use sea_orm::entity::prelude::*;
use sea_orm::Set;
use serial_test::serial;

// TODO: see how to dedup / extract this to app-local test utils
// not to framework, because that would require a runtime dep on insta
macro_rules! configure_insta {
    ($($expr:expr),*) => {
        let mut settings = insta::Settings::clone_current();
        settings.set_prepend_module_to_snapshot(false);
        settings.set_snapshot_suffix("notes_request");
        let _guard = settings.bind_to_scope();
    };
}

async fn authenticate_user(
    mut request: loco_rs::TestServer,
    email: &str,
    password: &str
) -> loco_rs::TestServer {
    let login_payload =
        serde_json::json!({
        "email": email,
        "password": password,
    });

    let login_response = request.post("/api/auth/login").json(&login_payload).await;

    assert_eq!(login_response.status_code(), 200);

    let body: serde_json::Value = serde_json::from_str(&login_response.text()).unwrap();
    let token = body["token"].as_str().unwrap();

    request.add_header(
        HeaderName::from_static("authorization"),
        HeaderValue::from_str(&format!("Bearer {}", token)).unwrap()
    );
    request
}

#[tokio::test]
#[serial]
async fn can_get_notes() {
    configure_insta!();

    testing::request::<App, _, _>(|request: loco_rs::TestServer, ctx| async move {
        testing::seed::<App>(&ctx.db).await.unwrap();
        let authenticated_request = authenticate_user(request, "edvinas1@gmail.com", "1234").await;

        let notes = authenticated_request.get("/api/notes").await;

        with_settings!({
            filters => {
                 let mut combined_filters = testing::CLEANUP_DATE.to_vec();
                    combined_filters.extend(vec![(r#"\"id\\":\d+"#, r#""id\":ID"#)]);
                    combined_filters
            }
        }, {
            assert_debug_snapshot!(
            (notes.status_code(), notes.text())
        );
        });
    }).await;
}

#[tokio::test]
#[serial]
async fn can_add_note() {
    configure_insta!();

    testing::request::<App, _, _>(|request, ctx| async move {
        testing::seed::<App>(&ctx.db).await.unwrap();

        let authenticated_request = authenticate_user(request, "edvinas1@gmail.com", "1234").await;

        let payload =
            serde_json::json!({
            "title": "loco",
            "content": "loco note test",
        });

        let add_note_request = authenticated_request.post("/api/notes").json(&payload).await;

        with_settings!({
            filters => {
                 let mut combined_filters = testing::CLEANUP_DATE.to_vec();
                    combined_filters.extend(vec![(r#"\"id\\":\d+"#, r#""id\":ID"#)]);
                    combined_filters
            }
        }, {
            assert_debug_snapshot!(
            (add_note_request.status_code(), add_note_request.text())
        );
        });
    }).await;
}

#[tokio::test]
#[serial]
async fn can_get_note() {
    configure_insta!();

    testing::request::<App, _, _>(|request, ctx| async move {
        testing::seed::<App>(&ctx.db).await.unwrap();

        let authenticated_request = authenticate_user(request, "edvinas1@gmail.com", "1234").await;

        let get_note_request = authenticated_request.get("/api/notes/3").await;

        with_settings!({
            filters => {
                 let mut combined_filters = testing::CLEANUP_DATE.to_vec();
                    combined_filters.extend(vec![(r#"\"id\\":\d+"#, r#""id\":ID"#)]);
                    combined_filters
            }
        }, {
            assert_debug_snapshot!(
            (get_note_request.status_code(), get_note_request.text())
        );
        });
    }).await;
}

#[tokio::test]
#[serial]
async fn can_delete_note() {
    configure_insta!();

    testing::request::<App, _, _>(|request, ctx| async move {
        testing::seed::<App>(&ctx.db).await.unwrap();

        let authenticated_request = authenticate_user(request, "edvinas1@gmail.com", "1234").await;
        let count_before_delete = Entity::find().all(&ctx.db).await.unwrap().len();
        let delete_note_request = authenticated_request.delete("/api/notes/3").await;

        with_settings!({
            filters => {
                 let mut combined_filters = testing::CLEANUP_DATE.to_vec();
                    combined_filters.extend(vec![(r#"\"id\\":\d+"#, r#""id\":ID"#)]);
                    combined_filters
            }
        }, {
            assert_debug_snapshot!(
            (delete_note_request.status_code(), delete_note_request.text())
        );
        });

        let count_after_delete = Entity::find().all(&ctx.db).await.unwrap().len();
        assert_eq!(count_after_delete, count_before_delete - 1);
    }).await;
}

#[tokio::test]
#[serial]
async fn can_share_note() {
    configure_insta!();

    testing::request::<App, _, _>(|request, ctx| async move {
        testing::seed::<App>(&ctx.db).await.unwrap();
        let authenticated_request = authenticate_user(request, "edvinas1@gmail.com", "1234").await;
        // find note by user
        let note_id = 3;

        let payload = serde_json::json!({
            "shared_with_user_id": 4
        });

        let share_note_request = authenticated_request
            .post(&format!("/api/notes/{}/share", note_id))
            .json(&payload).await;

        with_settings!({
            filters => {
                 let mut combined_filters = testing::CLEANUP_DATE.to_vec();
                 combined_filters.extend(vec![(r#"\"id\\":\d+"#, r#""id\":ID"#)]);
                 combined_filters
            }
        }, {
            assert_debug_snapshot!(
                (share_note_request.status_code(), share_note_request.text())
            );
        });

        // Verify the share was created
        let share = note_shares::Entity
            ::find()
            .filter(note_shares::Column::NoteId.eq(note_id))
            .one(&ctx.db).await
            .unwrap();
        assert!(share.is_some());
    }).await;
}

#[tokio::test]
#[serial]
async fn can_share_all_notes() {
    configure_insta!();

    testing::request::<App, _, _>(|request, ctx| async move {
        testing::seed::<App>(&ctx.db).await.unwrap();

        let authenticated_request = authenticate_user(request, "edvinas1@gmail.com", "1234").await;

        let payload = serde_json::json!({
            "shared_with_user_id": 4
        });

        // delete all notes shared with user 4
        note_shares::Entity
            ::delete_many()
            .filter(note_shares::Column::SharedWithUserId.eq(4))
            .exec(&ctx.db).await
            .unwrap();

        let share_all_notes_request = authenticated_request
            .post("/api/notes/share-all")
            .json(&payload).await;

        with_settings!(
            {},
            {
                assert_debug_snapshot!((
                    share_all_notes_request.status_code(),
                    share_all_notes_request.text(),
                ));
            }
        );

        // Verify that all notes are shared
        let shared_notes = note_shares::Entity
            ::find()
            .filter(note_shares::Column::SharedWithUserId.eq(4))
            .all(&ctx.db).await
            .unwrap();

        let user_notes = notes::Entity
            ::find()
            .filter(notes::Column::UserId.eq(3))
            .all(&ctx.db).await
            .unwrap();

        assert_eq!(shared_notes.len(), user_notes.len());
    }).await;
}

#[tokio::test]
#[serial]
async fn can_get_shared_notes() {
    configure_insta!();

    testing::request::<App, _, _>(|request, ctx| async move {
        testing::seed::<App>(&ctx.db).await.unwrap();
        let authenticated_request = authenticate_user(request, "edvinas1@gmail.com", "1234").await;
        let note = Entity::find().one(&ctx.db).await.unwrap().unwrap();

        let _share = (note_shares::ActiveModel {
            note_id: Set(note.id),
            shared_with_user_id: Set(2),
            ..Default::default()
        })
            .insert(&ctx.db).await
            .unwrap();

        let get_shared_notes_request = authenticated_request.get("/api/notes/shared").await;

        with_settings!({
            filters => {
                 let mut combined_filters = testing::CLEANUP_DATE.to_vec();
                 combined_filters.extend(vec![(r#"\"id\\":\d+"#, r#""id\":ID"#)]);
                 combined_filters
            }
        }, {
            assert_debug_snapshot!(
                (get_shared_notes_request.status_code(), get_shared_notes_request.text())
            );
        });
    }).await;
}

#[tokio::test]
#[serial]
async fn can_update_shared_note() {
    configure_insta!();

    testing::request::<App, _, _>(|request, ctx| async move {
        testing::seed::<App>(&ctx.db).await.unwrap();
        let authenticated_request = authenticate_user(request, "edvinas1@gmail.com", "1234").await;
        // Assume user 1 and 2 exist from seeding
        let note = Entity::find().one(&ctx.db).await.unwrap().unwrap();

        // Share the note
        let _share = (note_shares::ActiveModel {
            note_id: Set(note.id),
            shared_with_user_id: Set(4), // Assume user 4
            ..Default::default()
        })
            .insert(&ctx.db).await
            .unwrap();

        // Update the note
        let payload =
            serde_json::json!({
            "title": "Updated Shared Note",
            "content": "This note has been updated",
        });

        let update_note_request = authenticated_request
            .post(&format!("/api/notes/{}", note.id))
            .json(&payload).await;

        with_settings!({
            filters => {
                 let mut combined_filters = testing::CLEANUP_DATE.to_vec();
                 combined_filters.extend(vec![(r#"\"id\\":\d+"#, r#""id\":ID"#)]);
                 combined_filters
            }
        }, {
            assert_debug_snapshot!(
                (update_note_request.status_code(), update_note_request.text())
            );
        });

        // Verify the update
        let updated_note = Entity::find_by_id(note.id).one(&ctx.db).await.unwrap().unwrap();
        assert_eq!(updated_note.title, Some("Updated Shared Note".to_string()));
    }).await;
}
