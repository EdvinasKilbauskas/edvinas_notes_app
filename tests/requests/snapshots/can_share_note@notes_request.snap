---
source: tests/requests/notes.rs
expression: "(share_note_request.status_code(), share_note_request.text())"
---
(
    200,
    "{\"id\":ID,\"note_id\":3,\"shared_with_user_id\":4}",
)
